Add broadcasting support for `tf.where`

yongtang opened the pull request:
This is a rework on #15982

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add broadcasting support for `tf.where`

This is a rework on 15982

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add gradient of SelectV2

Credit to @brianwa84:

https://github.com/tensorflow/tensorflow/pull/15982#issuecomment-490226492

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang committed:
Add additional gradient tests for where_v2, also fixed a few issues

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

alextp commented on tensorflow/python/ops/math_grad.py:
```diff
@@ -1308,6 +1308,38 @@ def _SelectGrad(op, grad):
+  # Reduce away broadcasted leading dims.
+  gy = math_ops.reduce_sum(gy, axis=math_ops.range(rankdiff_y))
+  # Reduce but keep y's 1-valued dims which were broadcast.
+  axis = array_ops.where(gy_shape[rankdiff_y:] > y_shape)
```
Why is this where and not where_v2?

yongtang replied:
Thanks @alextp. The PR has been updated to use where_v2. In both cases (where_v2 and where), they behave the same though as `else` and `then` are `None` so follow the same rule. Updated to use `where_v2` for consistency.

alextp reviewed (changes requested):
We also need to have tests for the gradient code. This means at least a gradient checker test for the cases with and without broadcast. Running all existing where gradient tests with where_v2 as well is also a good idea.

This is core TF functionality and accidentally breaking it would be very bad, so it's worth the extra effort to get right.

yongtang committed:
Use array_ops.where_v2 in _SelectGradV2

The usages of array_ops.where_v2 and array_ops.where
are the same as else and then are None in both cases.
Use array_ops.where_v2 for consistency.

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
Thanks @alextp. The existing gradient tests also include where_v2 in `tensorflow/python/kernel_tests/cwise_ops_test.py`, the exiting non-gradient tests include where_v2 in   `tensorflow/python/kernel_tests/where_op_test.py`.

I will add additional test cases for broadcast that are special to where_v2 only. Will update again shortly.

yongtang committed:
Added additional gradient tests for where_v2

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@rthadur @alextp The failed tests are due to the api update. Should I update the api golden to make the tests pass?

yongtang committed:
Update API golden

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

yongtang commented:
@alexp @rthadur I updated the PR for api goldens. Believe all tests should pass now. Please let me know if there are any issues.

alextp commented on tensorflow/tools/api/golden/v2/tensorflow.pbtxt:
```diff
@@ -904,6 +904,10 @@ tf_module {
     argspec: "args=[\'sorted_sequence\', \'values\', \'side\', \'out_type\', \'name\'], varargs=None, keywords=None, defaults=[\'left\', \"<dtype: \'int32\'>\", \'None\'], "
   }
+  member_method {
+    name: "select_v2"
```
The selectv2 op needs an apidef file in the python_api directory to mark it as hidden

alextp reviewed (dismissed):
We also need to change the tf_upgrade_v2 script to rename tf.where_v2 to tf.compat.v2.where.

yongtang committed:
API compatibility update, and renames_v2 update

Signed-off-by: Yong Tang <yong.tang.github@outlook.com>

martinwicke commented:
It appears there is still something wrong:

```
import tensorflow.compat.v2 as tf
tf.enable_v2_behavior()

import numpy as np

print(np.where(tf.constant([[True]]), tf.ones([1, 10]), tf.zeros([1, 10])))
print(tf.where(tf.constant([[True]]), tf.ones([1, 10]), tf.zeros([1, 10])))

InvalidArgumentError: Inputs to operation Select of type Select must have the same size and shape.  Input 0: [1,1] != input 1: [1,10] [Op:Select]
```

yongtang commented:
@martinwicke Sorry about that. I will take a look and will fix it as soon as possible.

yongtang commented:
@martinwicke I tried with tf-nightly it looks fine:
```
# python
Python 2.7.15rc1 (default, Nov 12 2018, 14:31:15) 
[GCC 7.3.0] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import numpy as np   
>>> 
>>> import tensorflow.compat.v2 as tf
>>> tf.enable_v2_behavior()
>>> 
>>> import numpy as np
>>> 
>>> print(np.where(tf.constant([[True]]), tf.ones([1, 10]), tf.zeros([1, 10])))
2019-05-17 16:20:24.989573: I tensorflow/core/platform/cpu_feature_guard.cc:142] Your CPU supports instructions that this TensorFlow binary was not compiled to use: AVX2 AVX512F FMA
2019-05-17 16:20:25.050076: I tensorflow/core/platform/profile_utils/cpu_utils.cc:94] CPU Frequency: 2500000000 Hz
2019-05-17 16:20:25.050408: I tensorflow/compiler/xla/service/service.cc:168] XLA service 0x563b3ef74c80 executing computations on platform Host. Devices:
2019-05-17 16:20:25.050437: I tensorflow/compiler/xla/service/service.cc:175]   StreamExecutor device (0): <undefined>, <undefined>
[[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]]
>>> print(tf.where(tf.constant([[True]]), tf.ones([1, 10]), tf.zeros([1, 10])))
tf.Tensor([[1. 1. 1. 1. 1. 1. 1. 1. 1. 1.]], shape=(1, 10), dtype=float32)
>>> 
>>> print(tf.version.VERSION)
1.14.1-dev20190517
>>> print(tf.version.GIT_VERSION)
v1.12.1-2154-g3df6d99f3f
>>> 
```

Wondering if there is a version difference?

martinwicke commented:
Yeah, you are right. For some reason my nightly didn't pick this up yet, and I'm confused about that.

yongtang commented:
Thanks @martinwicke. Appreciate the help 👍

rthadur commented:
Changes are merged in to Master, closing this PR.
